sdk/webpubsub/azure-messaging-webpubsubservice/azure/messaging/webpubsubservice/_web_pub_sub_service_client.py

```diff
@@ -0,0 +1,209 @@
+        # type: (str, str, Any) -> Dict[Any]
+        """Build an authentication token for the given endpoint, hub using the provided key.
+
+        :keyword endpoint: connetion string or HTTP or HTTPS endpoint for the WebPubSub service instance.
+        :type endpoint: ~str
+        :keyword hub: The hub to give access to.
+        :type hub: ~str
+        :keyword accesskey: Key to sign the token with. Required if endpoint is not a connection string
+        :type accesskey: ~str
+        :keyword ttl: Optional ttl timedelta for the token. Default is 1 hour.
+        :type ttl: ~datetime.timedelta
+        :keyword user: Optional user name (subject) for the token. Default is no user.
+        :type user: ~str
+        :keyword roles: Roles for the token.
+        :type roles: typing.List[str]. Default is no roles.
+        :returns: ~dict containing the web socket endpoint, the token and a url with the generated access token.
+        :rtype: ~dict
+
+
+        Example:
+        >>> build_authentication_token(endpoint='https://contoso.com/api/webpubsub', hub='theHub', key='123')
+        {
+            'baseUrl': 'wss://contoso.com/api/webpubsub/client/hubs/theHub',
+            'token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ...',
+            'url': 'wss://contoso.com/api/webpubsub/client/hubs/theHub?access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ...'
+        }
+        """
+        if 'accesskey' not in kwargs:
+            kwargs = _parse_connection_string(endpoint, **kwargs)
+            endpoint = kwargs.pop('endpoint')
+
+        user = kwargs.pop("user", None)
+        key = kwargs.pop("accesskey")
+        ttl = kwargs.pop("ttl", timedelta(hours=1))
+        roles = kwargs.pop("roles", [])
+        endpoint = endpoint.lower()
+        if not endpoint.startswith("http://") and not endpoint.startswith("https://"):
+            raise ValueError(
+                "Invalid endpoint: '{}' has unknown scheme - expected 'http://' or 'https://'".format(
+                    endpoint
+                )
+            )
+
+        # Ensure endpoint has no trailing slash
+        endpoint = endpoint.rstrip("/")
+
+        # Switch from http(s) to ws(s) scheme
+        client_endpoint = "ws" + endpoint[4:]
+        client_url = "{}/client/hubs/{}".format(client_endpoint, hub)
+        audience = "{}/client/hubs/{}".format(endpoint, hub)
+
+        payload = {
+            "aud": audience,
+            "iat": datetime.now(tz=_UTC),
+            "exp": datetime.now(tz=_UTC) + ttl,
+        }
+        if user:
+            payload["sub"] = user
+        if roles:
+            payload["role"] = roles
+
+        token = six.ensure_str(jwt.encode(payload, key, algorithm="HS256"))
+        return {
+            "baseUrl": client_url,
+            "token": token,
+            "url": "{}?access_token={}".format(client_url, token),
+        }
+
+    @classmethod
+    def from_connection_string(cls, connection_string, **kwargs):
```

Comment:
Rather than modifying the generated code (and thus it being susceptible to being overwritten when regenerating), we should attempt to use the `patch_sdk` method in the `_patch.py` file to make modifications. Not needed for this release, however.

Reply:
I think it makes sense. But I can't find an example what should `patch_sdk()` looks like in [Azure/azure-sdk-for-python](https://github.com/Azure/azure-sdk-for-python)

Reply:
I add all the handwritten code into `patch_sdk()` to avoid changing generated code. Two problems left:
(1) `aio` does not have `_patch.py` and it will be deleted by codegen. I think we need to add it in autorest @iscai-msft 
(2) do not have better ways to change `__all__` in `__init__` (). So every time [these code](https://github.com/Azure/azure-sdk-for-python/blob/9cc207e88aee5466cb151bf3956f2ffc8b46256c/sdk/webpubsub/azure-messaging-webpubsubservice/azure/messaging/webpubsubservice/__init__.py#L19-L20) will be covered.